sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1882,3 +1882,311 @@ case class ArrayRepeat(left: Expression, right: Expression)
+              found = true
+            }
+            j = j + 1
+          }
+        }
+        if (!found) {
+          if (pos > MAX_ARRAY_LENGTH) {
+            throw new RuntimeException(s"Unsuccessful try to union arrays with $pos" +
+              s" elements due to exceeding the array size limit $MAX_ARRAY_LENGTH.")
+          }
+          array(pos) = v
+          pos = pos + 1
+        }
+      }))
+      new GenericArrayData(array.slice(0, pos))
+    }
+  }
+}
+
+abstract class ArraySetLike extends BinaryArrayExpressionWithImplicitCast {
+  def typeId: Int
+
+  override def dataType: DataType = left.dataType
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val typeCheckResult = super.checkInputDataTypes()
+    if (typeCheckResult.isSuccess) {
+      TypeUtils.checkForOrderingExpr(dataType.asInstanceOf[ArrayType].elementType,
+        s"function $prettyName")
+    } else {
+      typeCheckResult
+    }
+  }
+
+  private def cn = left.dataType.asInstanceOf[ArrayType].containsNull ||
+    right.dataType.asInstanceOf[ArrayType].containsNull
+
+  @transient private lazy val ordering: Ordering[Any] =
+    TypeUtils.getInterpretedOrdering(elementType)
+
+  @transient private lazy val elementTypeSupportEquals = elementType match {
+    case BinaryType => false
+    case _: AtomicType => true
+    case _ => false
+  }
+
+  def intEval(ary: ArrayData, hs2: OpenHashSet[Int]): OpenHashSet[Int]
+  def longEval(ary: ArrayData, hs2: OpenHashSet[Long]): OpenHashSet[Long]
+  def genericEval(ary: ArrayData, hs2: OpenHashSet[Any], et: DataType): OpenHashSet[Any]
+  def codeGen(ctx: CodegenContext, hs2: String, hs: String, len: String, getter: String, i: String,
+    postFix: String, newOpenHashSet: String): String
+
+  override def nullSafeEval(input1: Any, input2: Any): Any = {
+    val ary1 = input1.asInstanceOf[ArrayData]
+    val ary2 = input2.asInstanceOf[ArrayData]
+
+    if (!cn) {
+      elementType match {
+        case IntegerType =>
+          // avoid boxing of primitive int array elements
+          val hs2 = new OpenHashSet[Int]
+          var i = 0
+          while (i < ary2.numElements()) {
+            hs2.add(ary2.getInt(i))
+            i += 1
+          }
+          ArraySetLike.toArrayDataInt(intEval(ary1, hs2))
+        case LongType =>
+          // avoid boxing of primitive long array elements
+          val hs2 = new OpenHashSet[Long]
+          var i = 0
+          while (i < ary2.numElements()) {
+            hs2.add(ary2.getLong(i))
+            i += 1
+          }
+          ArraySetLike.toArrayDataLong(longEval(ary1, hs2))
+        case _ =>
+          val hs2 = new OpenHashSet[Any]
+          var i = 0
+          while (i < ary2.numElements()) {
+            hs2.add(ary2.get(i, elementType))
+            i += 1
+          }
+          new GenericArrayData(genericEval(ary1, hs2, elementType).iterator.toArray)
+      }
+    } else {
+      if (typeId == ArraySetLike.kindUnion) {
```

Comment:
How about moving to concrete class without using `typeId`?

Reply:
I eliminated `typeId` field by defining a method name in the concrete class.